Create Event class

jnnybk opened the pull request:
This PR sets up the Event Java class based on slide 21 in Data Structuring presentation: https://docs.google.com/presentation/d/1lGDiZT6umMahr-Jmqd8VLkIsO9dTW2PJkN3eCDV68PY/edit?usp=sharing.

The reason for choosing Map<Date, EmbeddedEntity> to represent the event's date(s) and hours was:
- an Event can span multiple days and have different hours for individual dates.

(EDIT) 

This PR creates Address class and Contact class. 

(EDIT 2)

Removed Address class and Contact class as separate classes based on pod's discussion with the interest of time + ease of use + simplicity of code

jnnybk committed:
Create Event class

jnnybk committed:
Create EventUpdater + AddEventServlet

belraquib commented on src/main/java/com/google/sps/data/Event.java:
```diff
@@ -0,0 +1,41 @@
+
+public class Event {
+
+  private long id;
```
comments for each field, please. I'm confused why some ids are strings and some are ints and some are embedded entities.

jnnybk committed:
Add comments explaining each field of Event

jnnybk committed:
delete irrelevant files

jnnybk committed:
change primitive to wrapper type

belraquib commented on src/main/java/com/google/sps/data/Event.java:
```diff
@@ -0,0 +1,43 @@
+
+public class Event {
+
+  private long id; // Represents ID of Event; Type is long because Datastore's Key's getId() return type is long.
```
nit: put the comment above the field, instead of next to.

belraquib commented on src/main/java/com/google/sps/data/Event.java:
```diff
@@ -0,0 +1,43 @@
+  private ArrayList<EmbeddedEntity> partnerIdsOrNames; // Holds EmbeddedEntity that represents an Organization's ID OR name. A partnering organization could have missing ID.
+  private String description; // Represents description of the Event. e.g. "In front of church ABC"
+  private Map<Date, EmbeddedEntity> dateAndHours; // Holds dates of Event and their corresponding time(s). 
+  private String contactEmail; // Represents a single contact email for this Event.
```
I wonder if there's a more generic Contact structure that can be used here and in the organizations as well. I can imagine contacts may also need names / terms of address?

jnnybk replied:
I will discuss this with the pod and update based on the discussion!

jnnybk replied:
After the pod discussion, we've agreed to keep it as it is because we don't feel the need to store address with the contact information.

I think that creating a contact class would be an option, too.

belraquib replied:
When I said terms of address I meant more the "I want to be called bob".

jnnybk replied:
Sorry! That makes a lot more sense. 

Would it be better to create a contact class having attributes for 
1) first name 
2) last name
3) phone number
4) email address,

or to have separate fields within the Event class holding this information?

belraquib replied:
The former (and you may want to do this for Organization contacts, if and when you have that as well)

jnnybk replied:
Okay, sounds good! Thank you!

I will make an issue to refactor Organization class since this pull request focuses on Event class! :)

belraquib commented on src/main/java/com/google/sps/data/Event.java:
```diff
@@ -0,0 +1,43 @@
+  private String contactPhone; // Represents a single contact phone number for this Event.
+
+  // Represents location information for the Event.
+  private String address;
```
Is there an opportunity to use a more generic Address structure for this and the Organizations?

belraquib replied:
Consider: https://cloud.google.com/appengine/docs/standard/python/datastore/typesandpropertyclasses#PostalAddress

belraquib replied:
java 8 version:
https://cloud.google.com/appengine/docs/standard/java/javadoc/com/google/appengine/api/datastore/PostalAddress

https://cloud.google.com/appengine/docs/standard/java/datastore/entity-property-reference

jnnybk replied:
Our pod's concern is that we would need to store the zipcode separately or somehow parse this (if all addresses were in the same format). We've considered using PostalAddress but it seems more appropriate to store these attributes separately.

belraquib replied:
That is valid - I hadn't realized it was unparsed! That all being said, consider creating your own Address class, so that there isn't a chance of the representation here skewing from the organization (or maybe you _do_ want that)

belraquib replied:
Can you do what you did for contact here?

Address {
   street_address
   city
   state
   zipcode
}

?

jnnybk replied:
Yes :)

jnnybk committed:
Move comments above each field

jnnybk committed:
Merge branch 'master' of https://github.com/googleinterns/step130-2020 into add-event-class

jnnybk committed:
Add time stamp attributes and contact name

jnnybk committed:
Add Contact class

jnnybk committed:
Add constructor for Event class

jnnybk committed:
Add address Java class

belraquib commented on src/main/java/com/google/sps/data/Address.java:
```diff
@@ -0,0 +1,31 @@
+
+public class Address {
+
+  private String address;
```
Add comments to each field that may be ambiguous or may need specifically formatted data (ie, only numbers). What do you expect in each? Eg. Is address the street + apt number or would that be in a separate field?

jnnybk replied:
I added these comments and an example!

belraquib commented on src/main/java/com/google/sps/data/Contact.java:
```diff
@@ -0,0 +1,36 @@
+package com.google.sps.data;
+
+public class Contact {
+  // Represents contact email.
```
Instead of repeating the field name, consider giving an example. 

Any email address, e.g.  help@theawesomeplace.com

belraquib commented on src/main/java/com/google/sps/data/Contact.java:
```diff
@@ -0,0 +1,36 @@
+public class Contact {
+  // Represents contact email.
+  private String email;
+  // Represents contact phone number.
```
Instead of/in addition to calling out that this is a phone number, tell the reader what to expect here:

A 10 digit US number: e.g. 111 111 1111 or the text equivalent, 800 HELP ALL1

?

belraquib commented on src/main/java/com/google/sps/data/Contact.java:
```diff
@@ -0,0 +1,36 @@
+  // Represents contact first name.
+  private String firstName;
+  // Represents contact last name.
+  private String lastName;
```
are both of these required? Can they be left empty? The comment is where you should specify your preconditions

belraquib reviewed (changes requested):
Getting there! I'm being very nitpicky, but the documentation is important, as it lets everyone get on the same page

jnnybk committed:
Merge branch 'master' of https://github.com/googleinterns/step130-2020 into add-event-class

jnnybk committed:
Add examples in comments

jnnybk committed:
Remove Address and Contact as separate classes

jnnybk commented:
UPDATE: Removed Address class and Contact class as separate classes based on pod's discussion with the interest of time.

jnnybk committed:
Fix import statements

jnnybk committed:
Update Event.java to have just ONE owner

Based on the frontend form design, this update was necessary.

jnnybk committed:
Update attribute

jnnybk committed:
change partnerIDsOrNames to partnerNames

partnerNames is represented as ArrayList of Strings: names of partnering organization.

jnnybk committed:
change attribute /property names
